File: apps/api-go/internal/validation/service.go
package validation

import (
	"api-go/internal/contracts"
)

func optionsSelectModeRule(v string) string {
	if v != string(contracts.OptionsSingle) && v != string(contracts.OptionsMulti) {
		return "Invalid input: expected one of single|multi"
	}
	return ""
}

func hasDuplicates(values []string) bool {
	seen := map[string]bool{}
	for _, v := range values {
		if seen[v] {
			return true
		}
		seen[v] = true
	}
	return false
}

// optionsConsistency — optionsSelectMode is required iff the service
// defines options. `null` counts as "no options": the cabinet clears an
// option list by sending options: null together with optionsSelectMode: null.
func optionsConsistency(e *Errors, options []string, optionsSelectMode *string) {
	hasOptions := len(options) > 0
	if hasOptions && optionsSelectMode == nil {
		e.Add("optionsSelectMode", "optionsSelectMode is required when options are set")
	}
	if !hasOptions && optionsSelectMode != nil {
		e.Add("optionsSelectMode", "optionsSelectMode must be omitted when there are no options")
	}
}

func optionalOptionsConsistency(e *Errors, options contracts.Optional[[]string], mode contracts.Optional[string]) {
	hasOptions := options.Set && options.Value != nil && len(*options.Value) > 0
	modeSet := mode.Set && mode.Value != nil
	if hasOptions && !modeSet {
		e.Add("optionsSelectMode", "optionsSelectMode is required when options are set")
	}
	if !hasOptions && modeSet {
		e.Add("optionsSelectMode", "optionsSelectMode must be omitted when there are no options")
	}
}

// ParseCreateServiceInput — port of createServiceInput.
func ParseCreateServiceInput(body []byte) (contracts.CreateServiceInput, *Errors) {
	m, e := rawObject(body)
	if e != nil {
		return contracts.CreateServiceInput{}, e
	}
	e = NewErrors()
	var out contracts.CreateServiceInput

	out.Title, _ = strValue(e, m, "title", true, true, DisplayNameRule)
	if v, present := strValue(e, m, "description", false, true, ServiceDescriptionRule); present {
		out.Description = &v
	}
	if v, present := strValue(e, m, "location", false, true, LocationRule); present {
		out.Location = &v
	}
	if v, present := strValue(e, m, "contact", false, true, ContactRule); present {
		out.Contact = &v
	}
	out.DefaultPrice, _ = strValue(e, m, "defaultPrice", true, true, PriceRule)
	capacity, _ := intValue(e, m, "defaultCapacity", true, intRange(1, 100_000))
	out.DefaultCapacity = int(capacity)
	duration, _ := intValue(e, m, "defaultDurationMinutes", true, intRange(1, 1440))
	out.DefaultDurationMinutes = int(duration)
	maxSeats, _ := intValue(e, m, "maxSeatsPerBooking", true, intRange(1, 1000))
	out.MaxSeatsPerBooking = int(maxSeats)

	options, _ := strArrValue(e, m, "options", false, true, OptionLabelRule, 50)
	if len(options) == 0 {
		options = nil // absent and empty both mean "no options"
	} else if len(options) > 0 && hasDuplicates(options) {
		e.Add("options", "options must be unique")
	}
	out.Options = options

	if v, present := strValue(e, m, "optionsSelectMode", false, false, optionsSelectModeRule); present {
		out.OptionsSelectMode = &v
	}
	if v, present := strValue(e, m, "photoUrl", false, false, URLRule); present {
		out.PhotoURL = &v
	}

	optionsConsistency(e, out.Options, out.OptionsSelectMode)
	return out, e.Finish()
}

// ParseUpdateServiceInput — cabinet edits: every optional display field
// is additionally nullable (absent = leave unchanged, null = clear).
func ParseUpdateServiceInput(body []byte) (contracts.UpdateServiceInput, *Errors) {
	m, e := rawObject(body)
	if e != nil {
		return contracts.UpdateServiceInput{}, e
	}
	e = NewErrors()
	var out contracts.UpdateServiceInput

	out.Title = optStr(e, m, "title", true, false, DisplayNameRule)
	out.Description = optStr(e, m, "description", true, true, ServiceDescriptionRule)
	out.Location = optStr(e, m, "location", true, true, LocationRule)
	out.Contact = optStr(e, m, "contact", true, true, ContactRule)
	out.DefaultPrice = optStr(e, m, "defaultPrice", true, false, PriceRule)
	out.DefaultCapacity = optInt(e, m, "defaultCapacity", false, intRange(1, 100_000))
	out.DefaultDurationMinutes = optInt(e, m, "defaultDurationMinutes", false, intRange(1, 1440))
	out.MaxSeatsPerBooking = optInt(e, m, "maxSeatsPerBooking", false, intRange(1, 1000))

	out.Options = optStrArr(e, m, "options", true, OptionLabelRule, 50)
	if out.Options.Set && out.Options.Value != nil {
		if len(*out.Options.Value) == 0 {
			out.Options.Value = nil // empty array clears like null
		} else if hasDuplicates(*out.Options.Value) {
			e.Add("options", "options must be unique")
		}
	}
	out.OptionsSelectMode = optStr(e, m, "optionsSelectMode", false, true, optionsSelectModeRule)
	out.PhotoURL = optStr(e, m, "photoUrl", false, true, URLRule)

	optionalOptionsConsistency(e, out.Options, out.OptionsSelectMode)
	return out, e.Finish()
}

File: packages/contracts/src/service.ts
import { z } from 'zod'

import { optionsSelectModeEnum } from './enums'
import { optionsList } from './options'
import {
  capacity,
  contact,
  displayName,
  durationMinutes,
  httpUrl,
  location,
  maxSeatsPerBooking,
  priceText,
  serviceDescription,
  serviceId,
  uuid,
} from './primitives'

const serviceFields = {
  title: displayName,
  description: serviceDescription.optional(),
  location: location.optional(),
  contact: contact.optional(),
  defaultPrice: priceText,
  defaultCapacity: capacity,
  defaultDurationMinutes: durationMinutes,
  maxSeatsPerBooking,
  options: optionsList.optional(),
  optionsSelectMode: optionsSelectModeEnum.optional(),
}

/**
 * `optionsSelectMode` is required iff the service defines options.
 * `null` counts as "no options": the cabinet clears an option list by sending
 * `options: null` together with `optionsSelectMode: null`.
 *
 * On the merge-patch (`PUT …/services/{id}`) this refinement runs on the
 * **patch**, not on the merged row — deliberately stricter than RFC 7386
 * (ADR-016): a patch that sets a non-empty `options` must carry
 * `optionsSelectMode` in the same document, and a mode without options is
 * rejected. The Go endpoint then validates the merged state as well
 * (`RefineServiceMergedState`), so a patch that clears `options` without
 * clearing the mode still answers 400 — the pair is one value split across
 * two columns and clients send it together (see `service-form.ts`).
 */
function refineOptionsConsistency<T extends z.ZodType>(schema: T) {
  return schema.superRefine((value, ctx) => {
    const { options, optionsSelectMode } = value as {
      options?: string[] | null
      optionsSelectMode?: string | null
    }
    const hasOptions = Array.isArray(options) && options.length > 0
    if (hasOptions && !optionsSelectMode) {
      ctx.addIssue({
        code: 'custom',
        path: ['optionsSelectMode'],
        message: 'optionsSelectMode is required when options are set',
      })
    }
    if (!hasOptions && optionsSelectMode) {
      ctx.addIssue({
        code: 'custom',
        path: ['optionsSelectMode'],
        message: 'optionsSelectMode must be omitted when there are no options',
      })
    }
  })
}

export const createServiceInput = refineOptionsConsistency(
  z.object({
    ...serviceFields,
    photoUrl: httpUrl.optional(),
  }),
)
export type CreateServiceInput = z.infer<typeof createServiceInput>

/**
 * Cabinet edits. Every optional display field is additionally **nullable**:
 * `undefined` means "leave unchanged", `null` means "clear it". Without that
 * distinction an organizer could never remove a description or a cover photo.
 */
export const updateServiceInput = refineOptionsConsistency(
  z
    .object({
      title: displayName,
      description: serviceDescription.nullable(),
      location: location.nullable(),
      contact: contact.nullable(),
      defaultPrice: priceText,
      defaultCapacity: capacity,
      defaultDurationMinutes: durationMinutes,
      maxSeatsPerBooking,
      options: optionsList.nullable(),
      optionsSelectMode: optionsSelectModeEnum.nullable(),
      photoUrl: httpUrl.nullable(), // null = remove cover photo
    })
    .partial(),
)
export type UpdateServiceInput = z.infer<typeof updateServiceInput>

/**
 * A service as returned by the API. Mirrors the `services` table with dates
 * normalized to ISO strings so it can cross the server/client boundary.
 */
export const serviceRecord = z.object({
  id: serviceId,
  organizerId: uuid,
  title: displayName,
  description: z.string().nullable(),
  photoUrl: z.string().nullable(),
  location: z.string().nullable(),
  contact: z.string().nullable(),
  defaultPrice: priceText,
  defaultCapacity: capacity,
  defaultDurationMinutes: durationMinutes,
  maxSeatsPerBooking,
  options: z.array(z.string()).nullable(),
  optionsSelectMode: optionsSelectModeEnum.nullable(),
  createdAt: z.string(),
})
export type ServiceRecord = z.infer<typeof serviceRecord>

/**
 * Display-field inheritance: a service may override its organizer's `location`
 * and `contact`, and falls back to the organizer's value when it has none.
 * Isomorphic on purpose — the public page, calendar link, and worker all need
 * this rule, so it cannot live in a component.
 */

interface OverridableDisplayFields {
  location?: string | null
  contact?: string | null
}

/** Location shown for a service: its own, else the organizer's. */
export function effectiveLocation(
  service: OverridableDisplayFields,
  organizer: OverridableDisplayFields,
): string | undefined {
  return service.location ?? organizer.location ?? undefined
}

/** Contact shown for a service: its own, else the organizer's. Same rule as location. */
export function effectiveContact(
  service: OverridableDisplayFields,
  organizer: OverridableDisplayFields,
): string | undefined {
  return service.contact ?? organizer.contact ?? undefined
}
